lib/_http_client.js

```diff
@@ -30,10 +30,17 @@ function ClientRequest(options, cb) {
 
   var agent = options.agent;
   var defaultAgent = options._defaultAgent || Agent.globalAgent;
+
+  if (agent != null && agent !== false && !(agent instanceof Agent.Agent)) {
+    throw new Error(
+       'Agent option must be an instance of http.Agent, undefined or false. ' +
+       `Received type: ${typeof agent} instead`
+    );
+  }
+
   if (agent === false) {
     agent = new defaultAgent.constructor();
-  } else if ((agent === null || agent === undefined) &&
-             typeof options.createConnection !== 'function') {
+  } else if (agent == null && typeof options.createConnection !== 'function') {
     agent = defaultAgent;
   }
```

Comment:
@cjihrig how about this logic structure:

```js
if (agent === false) {
   // noop
} else if (agent == null) {
   if (typeof options.createConnection !== 'function') {
     // noop
   }
} else if (!(agent instanceof Agent.Agent)) {
  // noop throw error.
}
```

if agent is undefined or null it'll be cause by the second logic gate, and if createConnection is a function it will not do anything, and will skip the final check. The last else if is exactly as you defined earlier. Obviously the code snippet has implementation removed just for brevity.